time circuit execution time and notify through ActiveSupport::Notifications

huangshan108 opened the pull request:
time circuit execution time and notify through `ActiveSupport::Notifications`

execution time notification will only be sent when the circuit is closed and execution successfully executed without timeout or errors.

msftclas commented:
@__huangshan108__,
Thanks for your contribution.
To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla.microsoft.com.

It will cover your contributions to all Microsoft-managed open source projects.
_Thanks,
Microsoft Pull Request Bot_

msftclas commented:
__@huangshan108__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, Microsoft Pull Request Bot_

sideshowcoder commented on test/circuit_breaker_test.rb:
```diff
@@ -403,19 +403,45 @@ def test_success_count_on_error_rate_calculation
+      notifier = gimme_notifier(metric: :execution_time, metric_value: Gimme::Matchers::Anything.new)
+      circuit = Circuitbox::CircuitBreaker.new(:yammer, notifier_class: notifier)
+      circuit.run { raise Timeout::Error }
+      assert !notifier.metric_sent?, 'execution time metric sent'
```
`assert not` is the same as `refute` might be a little easier to read.

huangshan108 commented:
@sideshowcoder Thanks for looking into. I have addressed your comments and added documentations. As well as bump the version to 2.1.0, if this number makex sense to you.

msftclas commented:
@huangshan108,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by Microsoft. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

sideshowcoder commented:
Thanks for working on this, seems really useful! I think @matthewshafer is probably the best person to get this out.

matthewshafer commented on test/circuit_breaker_test.rb:
```diff
@@ -403,19 +403,45 @@ def test_success_count_on_error_rate_calculation
+    def test_no_execution_time_metric_when_circuit_open
+      notifier = gimme_notifier(metric: :execution_time, metric_value: Gimme::Matchers::Anything.new)
+      circuit = Circuitbox::CircuitBreaker.new(:yammer, notifier_class: notifier)
+      10.times { circuit.run { raise Timeout::Error }}
```
Can we do this without the need to run the circuit 10 times, perhaps stubbing ```open?```. Then we would be able to test that when the circuit is open this doesn't run. It looks like this test is actually testing that we don't see any metrics sent when we see failures and then when the circuit is open, assuming it actually opened.

I see this is done in other tests in this file and I think we should change those separately.

matthewshafer commented on lib/circuitbox/circuit_breaker.rb:
```diff
@@ -186,6 +188,11 @@ def log_metrics(error_rate, failures, successes)
     end
 
+    def log_circuit_execution_time(time)
+      n = notifier.new(service)
```
I have a PR open to change this so we won't need to create a new notifier every time. So depending on when that gets merged this line might need some changes, just a heads up.

matthewshafer commented on test/circuit_breaker_test.rb:
```diff
@@ -403,19 +403,45 @@ def test_success_count_on_error_rate_calculation
       give(fake_notifier).notify_warning(Gimme::Matchers::Anything.new) { notified = true }
-      give(fake_notifier).metric_gauge(metric, metric_value) { notified = true }
+      give(fake_notifier).metric_gauge(metric, metric_value) do
+        notified = true
```
should clarify this a bit, I don't see it used in the tests added in this PR. I'm not sure if we use it elsewhere.

huangshan108 replied:
it's used in test `test_notifies_on_success_rate_calculation` and `test_success_count_on_error_rate_calculation`

matthewshafer commented on lib/circuitbox/circuit_breaker.rb:
```diff
@@ -52,15 +52,17 @@ def run!(run_options = {})
 
         begin
-          t1 = Time.now
+          start_time = Time.now.to_i
```
doing ```to_i``` will only give this to you in seconds so most things that happen under a second we'll either report a timing of 0 or 1.

matthewshafer commented on lib/circuitbox/circuit_breaker.rb:
```diff
@@ -52,15 +52,17 @@ def run!(run_options = {})
-          t2 = Time.now
-          log_circuit_execution_time(t2 - t1)
+          end_time = Time.now.to_i
+          if circuit_options[:notify_circuit_execution_time]
```
While I like this to allow users to disable notifications for execution time if it's disabled we still actually collect timings. I like the idea of letting users specify their own timer class/object (we can default to ```Time```.

My thought being if I had a time class like the one below I could pass an object of this class to my circuit:

```ruby
class MonotonicTime
  def initialize(time_unit: :millisecond)
    @time_unit = time_unit
  end

  def now
    Process.clock_gettime(Process::CLOCK_MONOTONIC, @time_unit)
  end
end
```

sideshowcoder replied:
Going that route doesn't it make more sense to actually define an interface which would not depend on time at all?

```ruby
class Timed

  def initialize(notifier, metric_name)
     @notifier = notifier
     @metric_name = metrics_name
   end

  def run
     before = Time.now
     yield
     after = Time.now
     notifier.metric_gauge(metric_name, before - after)
  end
end
```

This way we don't have any of the timing logic in here but can easily provide a `NullObject` which removes all the calls to stats.
Now this could be used where we need it and the only thing we expect is

matthewshafer replied:
The only thing I see us running into with that is when you want different time measurements you'll have to create a new class for each one since this class is initialized when the circuit is initialized.

Say for instance I wanted to use the monotonic clock in milliseconds for some work that was longer running and microseconds for one that was more latency sensitive. I might have something along the lines of
```ruby
class MonotonicTime
  def initialize(notifier, metric_name)
    @notifier = notifier
    @metric_name = metric_name
    @time_unit = :milliseconds
  end

 def run
    # pretty much the same logic here
  end
end

class MonotonicTimeMicroseconds < MonotonicTime
  def initialize(notifier, metric_name)
    super
    @time_unit = :microsecond
  end
end
```

I think we might be able to alleviate my concerns by doing something like this

```ruby
class Timed
  def run(notifier, metric_name)
    before = Time.now
    yield
    after = Time.now
    notifier.metric_gauge(metric_name, before - after)
  end
end
```
and then change the options fetching to
``` @time_class   = options.fetch(:timer) { Timed.new } ```
We could also change run to be a class method and then the timer can just be ```Timed```

This makes it possible for us to create a monotonic timer like this
```ruby
class MonotonicTime
  def initialize(time_unit)
    @time_unit = time_unit
  end

  def run(notifier, metric_name)
    before = Process.clock_gettime(Process::CLOCK_MONOTONIC, @time_unit)
    yield
    after = Process.clock_gettime(Process::CLOCK_MONOTONIC, @time_unit)
    notifier.metric_gauge(metric_name, before - after)
  end
end
```
and that can be initialized by the user when creating the circuit allowing them to specify the time unit.

thoughts?

sideshowcoder replied:
I think I would prefer to pass instances instead here this makes it easier to work with I find.

matthewshafer commented on test/circuit_breaker_test.rb:
```diff
@@ -403,19 +403,53 @@ def test_success_count_on_error_rate_calculation
+    def test_no_execution_time_metric_when_circuit_open
+      notifier = gimme_notifier(metric: :execution_time, metric_value: Gimme::Matchers::Anything.new)
+      circuit = Circuitbox::CircuitBreaker.new(:yammer, notifier_class: notifier, notify_circuit_execution_time: true)
+      circuit.send(:open!)
```
What if we stubbed ```open?```? I mostly want to see what your thoughts are on that vs what we're currently doing here.

huangshan108 replied:
If we stubbed `open?` We will need to stub all the sub-routine inside `open?` as well. Making the circuit actually open will do the trick for all. Either way we are just trying to control the code path and make sure the that test runs, and there is no harm to test the actually behavior anyway.

matthewshafer commented on lib/circuitbox/circuit_breaker.rb:
```diff
@@ -27,6 +27,7 @@ def initialize(service, options = {})
       @circuit_options = options
       @circuit_store   = options.fetch(:cache) { Circuitbox.circuit_store }
       @notifier        = options.fetch(:notifier_class) { Notifier }
+      @execution_timer = options.fetch(:execution_timer) { SimpleTimer.new }
```
can we put this behind a class method on ```Circuitbox```? something like ```Circuitbox.execution_timer```?

matthewshafer commented on lib/circuitbox/timer/null_timer.rb:
```diff
@@ -0,0 +1,9 @@
+class NullTimer
+  def initialize(time_unit = :milliseconds)
```
i'd probably move this to the monotonic timer because it's not used in null timer or simple timer. Then I don't think anything would need to inherit from NullTimer

matthewshafer commented on lib/circuitbox.rb:
```diff
@@ -7,6 +7,9 @@
 require 'circuitbox/notifier'
+require 'circuitbox/timer/null_timer'
+require 'circuitbox/timer/monotonic_timer'
+require 'circuitbox/timer/simple_timer'
```
Maybe we should only load the default one and then let users require the ones they need?

matthewshafer commented:
I think this looks fine for now. I'll probably make a change to the timers so it's easier for people to use the monotonic one.
